## Hedging classification

This notebooks will analyze hedging patterns in the collected articles corpus. 

1. Look at distributions, frequencies of hedge words in the documents
2. Train semisupervised classification algorithm on individual sentences


In [4]:
# Just text reading/cleaning
import pandas as pd
import numpy as np
from read_json import JsonHelpers
import analysis_helper_functions as helpers
import re
import json
import sys
import os
import clean_txt as clean
import nltk
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.naive_bayes import MultinomialNB, GaussianNB
from sklearn.svm import SVC




if sys.platform == "linux":
    studies_path = "/home/wiktor/git/MasterThesis/Literature/Competition_Studies_Database/studies.json"
    authors_path = "/home/wiktor/git/MasterThesis/Literature/Competition_Studies_Database/authors.json"
    studies_location = "/home/wiktor/Dropbox/Git/MasterThesis/Literature/Competition_Studies_Database/"
else:    
    studies_path = "/Users/Wiktor/Dropbox/Git/MasterThesis/Literature/Competition_Studies_Database/studies.json"
    authors_path = "/Users/Wiktor/Dropbox/Git/MasterThesis/Literature/Competition_Studies_Database/authors.json"
    studies_location = "/Users/Wiktor/Dropbox/Git/MasterThesis/Literature/Competition_Studies_Database/"

# Documents collection with labels
df_documents = helpers.read_txt_studies(studies_location)

# List of all documents as strings
documents = df_documents['document'].tolist()

# List of labels
article_labels = df_documents['label'].tolist()

# CB label
cb = JsonHelpers().central_bank_paper_label()

context_stopwords = ['panzar', 'rosse','panzarrosse', 'rossepanzar', 'prh','journal', 'sciencedirect',
                     'banking', 'rosse–panzar', 'vol', 'department', 'university', 'school', 'economics',
                     'business', 'email', 'cid', 'bank', 'banking', 'asset', 'assets', 'revenue', 'total', 
                     'hstatistic', 'h-statistic', 'competition', 'finance', 'banks']
# List of tokenized documents
texts_token_clean = [clean.normalize(text, context_stopwords, exclude_stops=False) for text in documents]

# List of clean strings
texts_str_clean = helpers.convert_list(texts_token_clean)


In [3]:
# Checking reported database
paper_database = df_documents['document'].apply(lambda x: re.findall(r"([^.]*?database[^.]*\.)", x.lower()))

KeyboardInterrupt: 

## Uncertanity index per paper

Using hedge words I calculate uncertanity index: number of hedge words in the document, divided by number of words in the document. 

In [5]:
# Open file with hedge words
hedge_words = []
with open('hedge_words.txt', 'rt') as file:
    for line in file:
        if '#' not in line:
            hedge_words.append(line.replace('\n', ''))
    
def uncertainty_index(documents, labels, hedge_words):
    """
    Calculate uncertanity index for all all collected papers
    
    number of hedge words seen / total number of words in text
    """

    hedge_score = []
    for document, label in zip(documents, labels):
        keys_count = [document.count(key) for key in hedge_words]
        hedge_score.append((label, sum(keys_count)/len(re.findall(r'\w+', document))))

    return hedge_score


# Undertanity index per document table
undertanity_df = pd.DataFrame(uncertainty_index(texts_str_clean, article_labels, hedge_words))
undertanity_df['cb'] = undertanity_df[0].isin(cb)
sorted_undertanity_df = undertanity_df.sort_values(by=1, ascending=False).reset_index(drop=True)

# Uncertanity Rank of CB papers 
uncertanity_df_cb = sorted_undertanity_df[sorted_undertanity_df['cb'] == True].reset_index()
uncertanity_df_cb.drop(['cb'], axis=1, inplace=True)
uncertanity_df_cb.columns = ['rank', 'article label', 'uncertanity score']
uncertanity_df_cb['rank'] = uncertanity_df_cb['rank'] + 1

# Saving table
uncertanity_df_cb.to_excel('uncertanity_score_cb.xlsx', index=False)

# Uncertanity Rank of NCB papers 
uncertanity_df_ncb = sorted_undertanity_df[sorted_undertanity_df['cb'] == False].reset_index()
uncertanity_df_ncb.drop(['cb'], axis=1, inplace=True)
uncertanity_df_ncb.columns = ['rank', 'article label', 'uncertanity score']
uncertanity_df_ncb['rank'] = uncertanity_df_ncb['rank'] + 1

# Saving table
uncertanity_df_ncb.to_excel('uncertanity_score_ncb.xlsx', index=False)

sorted_undertanity_df

0         1     cb
0   p032  0.089467   True
1   p048  0.082132  False
2   p074  0.077574  False
3   p053  0.073553   True
4   p070  0.068966  False
5   p003  0.066594  False
6   p057  0.065937  False
7   p034  0.064996  False
8   p014  0.063074  False
9   p051  0.061458  False
10  p073  0.060942  False
11  p005  0.060382  False
12  p019  0.059615  False
13  p065  0.058723   True
14  p054  0.058021   True
15  p071  0.057450  False
16  p001  0.057362   True
17  p033  0.057119  False
18  p064  0.056907  False
19  p024  0.056893   True
20  p045  0.056235  False
21  p020  0.055409  False
22  p010  0.054435  False
23  p030  0.054017  False
24  p012  0.053696  False
25  p062  0.053408  False
26  p046  0.053216  False
27  p044  0.052822  False
28  p068  0.052392  False
29  p063  0.051852  False
..   ...       ...    ...
37  p040  0.049898  False
38  p008  0.049850   True
39  p026  0.049391  False
40  p021  0.049267  False
41  p018  0.048830  False
42  p055  0.047481  False
43  p052  0.047365   True
44  p015  0.047349  False
45  p011  0.047323  False
46  p027  0.047280  False
47  p002  0.045825  False
48  p029  0.045794  False
49  p023  0.045343   True
50  p025  0.044986   True
51  p058  0.044118  False
52  p009  0.044008  False
53  p038  0.043304  False
54  p013  0.043248  False
55  p043  0.042368  False
56  p016  0.039640  False
57  p047  0.037596  False
58  p039  0.036176  False
59  p036  0.035672   True
60  p031  0.034222   True
61  p022  0.034218  False
62  p041  0.033079   True
63  p004  0.030500  False
64  p069  0.030435  False
65  p067  0.028608  False
66  p037  0.028284   True

[67 rows x 3 columns]

## Frequent used hedge words accross all papers and groups split

http://web.informatik.uni-mannheim.de/ponzetto/pubs/stajner17a

The following are considered hedge instances **(Medlock, Briscoe)**:
- data_Set to train: https://sraf.nd.edu/textual-analysis/resources/#LM%20Sentiment%20Word%20Lists (MacDonald word list) 


In [6]:
# Join all texts
single_corpa =  ' '.join(texts_str_clean)

# Hedge words distribution accross all texts
full_popular_hedge = []
for key in hedge_words:
    full_popular_hedge.append((key, single_corpa.count(key)))
    
# Data Frame with most popular hedge words
all_hedge = pd.DataFrame(full_popular_hedge).sort_values(by=1, ascending=False)


ncb_texts_clean = []
cb_texts_clean = []
for text, label in zip(texts_str_clean, article_labels):
    if label in cb:
        cb_texts_clean.append(text)
    else:
        ncb_texts_clean.append(text)
        
# Distributions of hedge words accross two groups of papers        
ncb_single_text_clean = ' '.join(ncb_texts_clean)
cb_single_text_clean = ' '.join(cb_texts_clean)

cb_popular_hedge = []
ncb_popular_hedge = []

for key in hedge_words:
    cb_popular_hedge.append((key, cb_single_text_clean.count(key), cb_single_text_clean.count(key)/len(re.findall(r'\w+', cb_single_text_clean))))
    ncb_popular_hedge.append((key, ncb_single_text_clean.count(key), ncb_single_text_clean.count(key)/len(re.findall(r'\w+', ncb_single_text_clean))))
    
cb_hedge = pd.DataFrame(cb_popular_hedge).sort_values(by=1, ascending=False)
ncb_hedge = pd.DataFrame(ncb_popular_hedge).sort_values(by=1, ascending=False)



In [11]:
# All papers popular hedge
all_hedge

# CB papers popular hedge
cb_hedge

# NCB papers popular hedge
ncb_hedge

0    1         2
100               differ  780  0.004046
86                depend  660  0.003424
160             indicate  531  0.002755
277                 risk  482  0.002500
298                 some  428  0.002220
321              suggest  409  0.002122
118                 find  391  0.002028
180                 most  343  0.001779
72              consider  330  0.001712
175                 many  246  0.001276
420                would  238  0.001235
168                 like  223  0.001157
42                assume  210  0.001089
81                 could  181  0.000939
332                 tend  179  0.000929
409            variation  177  0.000918
46            assumption  159  0.000825
289                 seem  157  0.000814
178                might  144  0.000747
203            potential  140  0.000726
249                 read  137  0.000711
201             possible  137  0.000711
28                appear  133  0.000690
194              overall  131  0.000680
248               rather  128  0.000664
295              several  123  0.000638
239               random  102  0.000529
182                 much   99  0.000514
413                 vary   94  0.000488
170               likely   94  0.000488
..                   ...  ...       ...
173           looks like    0  0.000000
179         more or less    0  0.000000
31             appearing    0  0.000000
29          appear to be    0  0.000000
184        my impression    0  0.000000
185       my thinking is    0  0.000000
186  my understanding is    0  0.000000
189        nonassessable    0  0.000000
190           obtainable    0  0.000000
25         anticipations    0  0.000000
198            plausibly    0  0.000000
23          anticipating    0  0.000000
22           anticipates    0  0.000000
208        precautionary    0  0.000000
231        probabilistic    0  0.000000
209          precautions    0  0.000000
1               abeyance    0  0.000000
211       predictability    0  0.000000
213           predicting    0  0.000000
216            predictor    0  0.000000
217           predictors    0  0.000000
218             predicts    0  0.000000
219        preliminarily    0  0.000000
221           presumable    0  0.000000
224             presumed    0  0.000000
225             presumes    0  0.000000
226            presuming    0  0.000000
228         presumptions    0  0.000000
229       presumptuously    0  0.000000
0                 
a bit    0  0.000000

[421 rows x 3 columns]

In [7]:
# Hypothesis testing words
testing_words = []

full_popular_testing = []
cb_popular_testing = []
ncb_popular_testing = []

with open('hypothesis_testing_words.txt', 'rt') as file:
    for line in file:
        if '#' not in line:
            testing_words.append(line.replace('\n', ''))

for key in testing_words:
    cb_popular_testing.append((key, cb_single_text_clean.count(key), cb_single_text_clean.count(key)/len(re.findall(r'\w+', cb_single_text_clean))))
    ncb_popular_testing.append((key, ncb_single_text_clean.count(key), ncb_single_text_clean.count(key)/len(re.findall(r'\w+',ncb_single_text_clean))))
    full_popular_testing.append((key, single_corpa.count(key), single_corpa.count(key)/len(re.findall(r'\w+',single_corpa))))


cb_testing = pd.DataFrame(cb_popular_testing).sort_values(by=2, ascending=False)
ncb_testing = pd.DataFrame(ncb_popular_testing).sort_values(by=2, ascending=False)
all_testing = pd.DataFrame(full_popular_testing).sort_values(by=2, ascending=False)



In [13]:
# All papers popular hedge
all_testing

# CB papers popular hedge
cb_testing

# NCB papers popular hedge
ncb_testing

0    1         2
2                   reject  447  0.002319
3               hypothesis  376  0.001951
7              significant  345  0.001790
19                   error  213  0.001105
6              alternative  122  0.000633
4          null hypothesis  105  0.000545
15                  pvalue   89  0.000462
8            significantly   84  0.000436
9       significance level   38  0.000197
1              reject null   37  0.000192
11     confidence interval    6  0.000031
10        confidence level    4  0.000021
12    confidence intervals    1  0.000005
13             lower bound    0  0.000000
14             upper bound    0  0.000000
5   alternative hypothesis    0  0.000000
16                 p-value    0  0.000000
17            type i error    0  0.000000
18           type ii error    0  0.000000
0           fail to reject    0  0.000000

### Sentence uncertanity/vagueness analysis

To train supervised classifiers, we first transformed the sentences from the articles into a vector of numbers. We explored vector representations such as TF-IDF weighted vectors.

After having this vector representations of the text we can train supervised classifiers to predict the “uncertanity” of the sentence.

http://web.informatik.uni-mannheim.de/ponzetto/pubs/stajner17a

The following are considered hedge instances **(Medlock, Briscoe)**:
- data_Set to train: https://sraf.nd.edu/textual-analysis/resources/#LM%20Sentiment%20Word%20Lists (MacDonald word list) 


The Wikipedia training set (CoNLL-2010 shared task, v2.0,
task1) was used as the training set for all our experiments,
as it is the largest existing general-domain dataset annotated
for speculation. It contains a total of 11111 sentences, out of
which 2346 were marked as speculative (uncertain). A few
examples from this dataset are presented in Table V.

use :  Task1 wikipedia other variant (sentence tags in separate lines)

T. Loughran and B. McDonald, “When is a Liability not a Liability?
Textual Analysis, Dictionaries, and 10-Ks,” Journal of Finance, 2010.


The feature
set consisted of frequencies of each speculation trigger plus an additional feature (the total number of triggers found).

In [8]:
# Create df with text and corresponding label
sentence_df = pd.concat([pd.Series(texts_str_clean), pd.Series(article_labels)], ignore_index=True, axis=1)

# Label columns
sentence_df.columns = ['sentence', 'label']

# Expand columns by splitting text into sentences on '.'
sentence_df = sentence_df.set_index(sentence_df.columns.drop('sentence',1).tolist())\
                                                                          .sentence.str.split('.', expand=True)\
                                                                          .stack()\
                                                                          .reset_index()\
                                                                          .rename(columns={0:'sentence'})\
                                                                          .loc[:, sentence_df.columns]

# Drop sentences with les than 20 characters
sentence_df = sentence_df[sentence_df['sentence'].apply(lambda x: len(x)>20)]

# Count number of words per sentence
sentence_df['no_words'] = sentence_df['sentence'].apply(lambda x: len(re.findall(r'\w+', x)))

# Assign "uncertanity category"
sentence_df['uncertain'] = sentence_df['sentence'].str.contains('|'.join(hedge_words))

# Convert boolean to integer
sentence_df['uncertain'] = sentence_df['uncertain'].apply(lambda x: x*1)

sentence_df['cb'] = sentence_df['label'].isin(cb)

In [10]:
# TFIDF setup
tfidf = TfidfVectorizer(ngram_range=(1, 3), stop_words='english')

X_tfidf = tfidf.fit_transform(sentence_df['sentence'])


X_train, X_test, y_train, y_test = train_test_split(X_tfidf,
                                                    sentence_df['uncertain'], random_state = 0, test_size=0.3)

# Random forest with 100 trees
clf = RandomForestClassifier(n_estimators=100, n_jobs=-1).fit(X_train, y_train)

In [16]:
confusion_matrix(clf.predict(X_test), y_test)

print(classification_report(y_test, clf.predict(X_test)))

             precision    recall  f1-score   support

          0       0.80      0.98      0.88      2673
          1       0.97      0.72      0.83      2396

avg / total       0.88      0.86      0.85      5069



In [17]:
pd.DataFrame(confusion_matrix(y_test, clf.predict(X_test)))

0     1
0  2623    50
1   674  1722

In [18]:
# tfidf_list = []

# for i in X_tfidf:
#     tfidf_list.append(clf.predict(i)[0])
    
# # Make full sample (article) prediction, classifier tesed on the sample
# sentence_df['prediction_on_sample'] = tfidf_list

In [19]:
sentence_cb = sentence_df[sentence_df['label'].isin(cb)]
sentence_ncb = sentence_df[~sentence_df['label'].isin(cb)]

## Prediction using wikipedia sentence data

In [13]:
# Read data, 22222 sentences
wiki_data = pd.read_csv('wiki_data_clean.csv')

# Transform string certanity variable into dummy 
wiki_data['uncertanity_dummy'] = wiki_data['certanity'].str.get_dummies()['uncertain']

# Dropping sentences that were NaN (not string)
wiki_data = wiki_data[wiki_data['sentence'].apply(lambda x: isinstance(x, str))]

# Selecting only strings with more than 20 characters (removing single words), 18744 remain
wiki_data = wiki_data[wiki_data['sentence'].apply(lambda x: len(x) > 20)]

# Transform data into TFIDF, adding data from the articles to have appropreate number of features 
# first 18744 rows are wiki, remainder articles
wiki_tfidf = tfidf.fit_transform(pd.concat([wiki_data['sentence'], sentence_df['sentence']]))

In [21]:
# Training random forest on wikipedia data
rf_wiki = RandomForestClassifier(n_estimators=100, n_jobs=-1).fit(wiki_tfidf[:18744],
                                                                   wiki_data['uncertanity_dummy'])

nb_wiki = MultinomialNB().fit(wiki_tfidf[:18744],
                              wiki_data['uncertanity_dummy'])

svc_wiki = SVC(kernel='linear').fit(wiki_tfidf[:18744],
                     wiki_data['uncertanity_dummy'])

# Predict uncertanity based on article data using RF
rf_wiki_prediction = rf_wiki.predict(wiki_tfidf[18744:])

# Predict uncertanity based on article data using NB
nb_wiki_prediction = nb_wiki.predict(wiki_tfidf[18744:])

# Predict uncertanity based on article data using SVM
svc_wiki_prediction = svc_wiki.predict(wiki_tfidf[18744:])


In [31]:
# Problems: unsupported words, seen in article but not in wiki
# no proper way to measure the prediction

sentence_df['rf_wiki_predicion'] = rf_wiki_prediction 
sentence_df['nb_wiki_predicion'] = nb_wiki_prediction 
sentence_df['svc_wiki_prediction'] = svc_wiki_prediction 



# Calculation fraction of sentences considered to be uncertain within each group RF
sentence_df[sentence_df['label'].isin(cb)]['rf_wiki_predicion'].sum()/len(sentence_df[sentence_df['label'].isin(cb)]['rf_wiki_predicion'])
sentence_df[~sentence_df['label'].isin(cb)]['rf_wiki_predicion'].sum()/len(sentence_df[~sentence_df['label'].isin(cb)]['rf_wiki_predicion'])

# Calculation fraction of sentences considered to be uncertain within each group SVM
sentence_df[sentence_df['label'].isin(cb)]['svc_wiki_prediction'].sum()/len(sentence_df[sentence_df['label'].isin(cb)]['svc_wiki_prediction'])
sentence_df[~sentence_df['label'].isin(cb)]['svc_wiki_prediction'].sum()/len(sentence_df[~sentence_df['label'].isin(cb)]['svc_wiki_prediction'])


0.030495654972642422

1. RF classified 508 as uncertain out of 16903
    - 0.0281 of all CB sentences are considered to be uncertain
    - 0.0305 of all NCB sentence are considered to be uncertain

2. SVM classified 526 as uncertain out of 16903
    - 0.0372 of all CB sentences are considered to be uncertain
    - 0.0297 of all NCB sentence are considered to be uncertain

e.g "The estimation results for Specification II lead to different and
possibly misleading inferences about the nature of competition, if
they are interpreted using the same criteria regardless of the inclusion
of ln(yi,t) as an additional control variable in Specification II."

CB:
"Our results suggest that
European banks were operating under conditions of monopolistic competition and
that bank interest revenues in the 10 new EU member states was earned under
conditions of higher competition than those that existed in the old EU banking
countries"


In [23]:
# Creating ranking of papers where share of uncertain sentences is the highest

uncertain_sentences_count = sentence_df.groupby('label').agg({'svc_wiki_prediction': 'sum',
                                                              'sentence': 'count'}).reset_index()

uncertain_sentences_count['uncertan_sentence_index'] = uncertain_sentences_count['svc_wiki_prediction']/uncertain_sentences_count['sentence']
uncertain_sentences_count['cb'] = uncertain_sentences_count['label'].isin(cb)

# Ranking of papers with highest fraction of uncertain sentences
uncertain_sentences_count.sort_values('uncertan_sentence_index', ascending=False).reset_index(drop=True)

label  svc_wiki_prediction  sentence  uncertan_sentence_index     cb
0   p026                   15       174                 0.086207  False
1   p032                   25       302                 0.082781   True
2   p059                    5        80                 0.062500   True
3   p048                   16       260                 0.061538  False
4   p074                   20       344                 0.058140  False
5   p062                   15       259                 0.057915  False
6   p034                   10       181                 0.055249  False
7   p057                   10       183                 0.054645  False
8   p065                   19       355                 0.053521   True
9   p051                    6       114                 0.052632  False
10  p007                    9       175                 0.051429  False
11  p029                    6       117                 0.051282  False
12  p011                   18       355                 0.050704  False
13  p020                    5       100                 0.050000  False
14  p009                    9       200                 0.045000  False
15  p045                   10       225                 0.044444  False
16  p067                    7       158                 0.044304  False
17  p027                    7       171                 0.040936  False
18  p018                    5       126                 0.039683  False
19  p053                   14       358                 0.039106   True
20  p063                   13       337                 0.038576  False
21  p069                    4       108                 0.037037  False
22  p003                    8       235                 0.034043  False
23  p023                    7       207                 0.033816   True
24  p042                    6       181                 0.033149  False
25  p041                    4       122                 0.032787   True
26  p068                    9       284                 0.031690  False
27  p031                    8       256                 0.031250   True
28  p061                    7       227                 0.030837  False
29  p073                   10       326                 0.030675  False
..   ...                  ...       ...                      ...    ...
37  p010                    9       346                 0.026012  False
38  p025                    7       275                 0.025455   True
39  p054                    6       238                 0.025210   True
40  p016                    5       201                 0.024876  False
41  p015                    5       203                 0.024631  False
42  p038                    3       122                 0.024590  False
43  p033                   13       533                 0.024390  False
44  p008                   10       419                 0.023866   True
45  p066                    5       213                 0.023474  False
46  p024                    6       259                 0.023166   True
47  p002                    7       313                 0.022364  False
48  p044                    5       226                 0.022124  False
49  p013                    3       136                 0.022059  False
50  p036                   17       772                 0.022021   True
51  p030                    4       182                 0.021978  False
52  p012                    6       278                 0.021583  False
53  p005                    5       233                 0.021459  False
54  p052                    4       187                 0.021390   True
55  p070                    5       257                 0.019455  False
56  p022                    4       212                 0.018868  False
57  p040                    4       229                 0.017467  False
58  p021                    3       177                 0.016949  False
59  p043                    4       246                 0.016260  False
60  p019                    3       227    

In [24]:
test = uncertain_sentences_count.sort_values('uncertan_sentence_index', ascending=False).reset_index(drop=True)

test[test.cb == False].reset_index().to_excel('sentence_uncertainity_score_ncb.xlsx', index=False)

In [37]:
# Arevage number of words in certain vs uncertain sentence
sentence_df.groupby(['svc_wiki_prediction']).agg({'sentence': 'count',
                                                  'no_words': 'mean'}).reset_index()

# With CB split
sentence_df.groupby(['svc_wiki_prediction', 'cb']).agg({'sentence': 'count',
                                                   'no_words': 'mean'}).reset_index()

svc_wiki_prediction     cb  sentence   no_words
0                    0  False     12049  15.287825
1                    0   True      4319  16.317898
2                    1  False       379  16.947230
3                    1   True       147  19.884354

In [26]:
# Average number of words certain vs uncertain sentence per paper

test = sentence_df.groupby(['label', 'svc_wiki_prediction']).agg({'sentence': 'count',
                                                                  'no_words': 'mean'}).reset_index()

# Top 10 articles with highest uncertain share
test10 = test[test.label.isin(['p026', 'p032', 'p059', 'p048', 'p074', 'p062', 'p034', 'p057', 'p065', 'p051', 'p007'])]


# Top articles with highest uncertain share have average number of words in unc = 18.586
# and uncer = 15.74275851992153
test10.loc[test10['svc_wiki_prediction'] == 0, 'no_words'].mean()


label  svc_wiki_prediction  sentence   no_words
10   p007                    0       166  14.722892
11   p007                    1         9  21.000000
46   p026                    0       159  16.446541
47   p026                    1        15  19.400000
56   p032                    0       277  18.429603
57   p032                    1        25  26.880000
60   p034                    0       171  12.807018
61   p034                    1        10  14.200000
88   p048                    0       244  14.643443
89   p048                    1        16  15.625000
90   p051                    0       108  18.157407
91   p051                    1         6  21.333333
100  p057                    0       173  17.208092
101  p057                    1        10  13.500000
104  p059                    0        75  14.160000
105  p059                    1         5  17.400000
108  p062                    0       244  14.000000
109  p062                    1        15  19.533333
114  p065                    0       336  16.175595
115  p065                    1        19  20.526316
132  p074                    0       324  16.419753
133  p074                    1        20  15.050000

In [24]:
# Validating Random Forest vs SVM

y = wiki_data['uncertanity_dummy']
wiki_tfidf = tfidf.fit_transform(wiki_data['sentence'])
x = wiki_tfidf

X_train, X_test, y_train, y_test = train_test_split(x, y, random_state = 0, test_size=0.3)

# Training random forest on wikipedia data
rf_wiki = RandomForestClassifier(n_estimators=50, n_jobs=-1).fit(X_train, y_train)

svc_wiki = SVC(kernel='linear').fit(X_train, y_train)

# Predict uncertanity based on article data using RF
rf_wiki_prediction = rf_wiki.predict(X_test)

# Predict uncertanity based on article data using SVM
svc_wiki_prediction = svc_wiki.predict(X_test)

print(classification_report(y_test, rf_wiki_prediction))
print(classification_report(y_test, svc_wiki_prediction))


             precision    recall  f1-score   support

          0       0.89      1.00      0.94      4094
          1       0.98      0.68      0.81      1530

avg / total       0.92      0.91      0.90      5624

             precision    recall  f1-score   support

          0       0.89      1.00      0.94      4094
          1       0.99      0.67      0.80      1530

avg / total       0.92      0.91      0.90      5624



In [36]:
# PRecision
(0.91* 4094 + 0.74*1530)/5624

# Recall
#(0.98* 4094 + 0.62*1530)/5624


0.8637517780938834

In [22]:
pd.DataFrame(confusion_matrix(y_test, rf_wiki_prediction))

0     1
0  4080    14
1   499  1031

In [23]:
pd.DataFrame(confusion_matrix(y_test, svc_wiki_prediction))

0     1
0  4080    14
1   503  1027